src/main/java/com/silzila/querybuilder/RelativeCalendarFilter.java

```diff
@@ -0,0 +1,70 @@
+package com.silzila.querybuilder;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.h2.expression.Alias;
+
+import com.silzila.helper.AilasMaker;
+import com.silzila.payload.request.Dimension;
+import com.silzila.payload.request.Measure;
+import com.silzila.payload.request.Query;
+
+public class RelativeCalendarFilter {
+
+    public static String buildRelativeCalendarFilter(Query req, String CTEquery, String relativeType) {
+
+        Map<String, Integer> aliasNumbering = new HashMap<>();
+
+        String CTEqueryString = "WITH CTE AS (\n" + CTEquery + "\n) \n";
+
+        List<Dimension> dimensions = req.getDimensions();
+
+        List<Measure> measures = req.getMeasures();
+
+        List<String> selectWindowClause = new ArrayList<>();
+
+        for (Measure measure : measures) {
+
+            String alias = AilasMaker.aliasing(measure.getFieldName(), aliasNumbering);
+            String field = measure.getAggr().name() + "(" + alias
+                    + ") over (" + partitionBy(dimensions) + ")";
+            selectWindowClause.add(field);
+        }
+
+        StringBuilder finalRelativeQuery = new StringBuilder();
+        finalRelativeQuery.append(CTEqueryString);
+        finalRelativeQuery.append("Select * , ");
+        finalRelativeQuery.append(String.join(", ", selectWindowClause));
+        finalRelativeQuery.append("\n from CTE");
+
+        return finalRelativeQuery.toString();
+    }
+
+    public static String partitionBy(List<Dimension> dimensions) {
+
+        String partitionQuery = "";
+        List<Dimension> DimensionWithoutDate = new ArrayList<>();
+        for (int i = 0; i < dimensions.size(); i++) {
```

Comment:
```suggestion
        for (int i = 0; i < dimensions.size(); i++) {
            Dimension dim = dimensions.get(i);
            List<String> partitionClause = new ArrayList();
            if (!("date".equals(dim.getDataType().toString()) || "timestamp".equals(dim.getDataType().toString()))) {
                partitionClause.add(dim.getFieldName())
            }
        }
        if (partitionClause.size() > 0) {
            partitionQuery = " partition by " + String.join(', ', partitionClause);
        }
        return partitionQuery
    
```

You are doing 2 for loop to compute partition, but it can be simplified into one. like the one I suggested above.
This saves some time. I understand that the dimensions are not going to be huge so we can live with what you have as well. Only thing that looks kinda not good is finding the last index so that you can avoid appending comma to it. There is an inbuilt library join that helps you achieve what you wanted. So always try not to implement a code that can be easily done by libraries.